xcm/xcm-simulator/example/src/lib.rs

```diff
@@ -182,33 +192,201 @@ mod tests {
 			));
 		});
 
-		ParaB::execute_with(|| {
+		ParaB::execute_and_dispatch_xcm(|| {
 			use parachain::{Event, System};
 			assert!(System::events()
 				.iter()
 				.any(|r| matches!(r.event, Event::System(frame_system::Event::Remarked(_, _)))));
 		});
 	}
 
+	/// Scenario:
+	/// A user Alice sends funds from the relaychain to a parachain.
+	///
+	/// Asserts that the correct XCM is sent and the balances are set as expected.
 	#[test]
-	fn reserve_transfer() {
+	fn reserve_transfer_assets() {
 		MockNet::reset();
 
-		Relay::execute_with(|| {
+		let withdraw_amount = 123;
+		let max_weight_for_execution = 10;
+
+		Relay::execute_and_dispatch_xcm(|| {
 			assert_ok!(RelayChainPalletXcm::reserve_transfer_assets(
 				relay_chain::Origin::signed(ALICE),
 				X1(Parachain(1)),
 				X1(Junction::AccountId32 { network: NetworkId::Any, id: ALICE.into() }),
-				vec![ConcreteFungible { id: Null, amount: 123 }],
-				123,
+				vec![ConcreteFungible { id: Null, amount: withdraw_amount }],
+				max_weight_for_execution,
 			));
+			let para_account_a = ParaId::from(1).into_account();
+			assert_eq!(
+				parachain::Balances::free_balance(&para_account_a),
+				INITIAL_BALANCE + withdraw_amount
+			);
 		});
 
-		ParaA::execute_with(|| {
-			// free execution, full amount received
+		ParaA::execute_and_dispatch_xcm(|| {
+			// Check message received
+			let expected_message = (
+				X1(Parent),
+				ReserveAssetDeposit {
+					assets: vec![ConcreteFungible { id: X1(Parent), amount: withdraw_amount }],
+					effects: vec![
+						buy_execution(max_weight_for_execution),
+						Order::DepositAsset {
+							assets: vec![All],
+							dest: X1(Junction::AccountId32 {
+								network: NetworkId::Any,
+								id: ALICE.into(),
+							}),
+						},
+					],
+				},
+			);
+			assert_eq!(parachain::MsgQueue::received_dmp(), vec![expected_message]);
+			// Check message execution with full amount received
 			assert_eq!(
 				pallet_balances::Pallet::<parachain::Runtime>::free_balance(&ALICE),
-				INITIAL_BALANCE + 123
+				INITIAL_BALANCE + withdraw_amount
+			);
+		});
+	}
+
+	/// Scenario:
+	/// A parachain transfers funds on the relay chain to another parachain account.
+	///
+	/// Asserts that the parachain accounts are updated as expected.
+	#[test]
+	fn withdraw_and_deposit() {
+		MockNet::reset();
+
+		let send_amount = 10;
+		let mut amount_received = 0;
+		let weight_for_execution = 3 * relay_chain::BaseXcmWeight::get();
+
+		ParaA::execute_and_dispatch_xcm(|| {
+			let message = WithdrawAsset {
+				assets: vec![ConcreteFungible { id: Null, amount: send_amount }],
+				effects: vec![
+					buy_execution(weight_for_execution),
+					Order::DepositAsset { assets: vec![All], dest: Parachain(2).into() },
+				],
+			};
+			// Send withdraw and deposit
+			assert_ok!(ParachainPalletXcm::send_xcm(Null, X1(Parent), message.clone(),));
+		});
+
+		amount_received += send_amount;
```

Comment:
this is superfluous in the case of 1 transfer. You initialize it, increment it by `send_amount` and then check.  This should go away and the last line would be: 

```
assert_eq!(relay_chain::Balances::free_balance(para_account_b), send_amount);
```

```suggestion
```